src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SparseIntMap.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Text.RegularExpressions.Symbolic
+{
+    /// <summary>An insertion-ordered map that supports small int keys.</summary>
+    /// <remarks>Uses a sparse array of the same size as the space of keys for efficient lookups.</remarks>
+    internal sealed class SparseIntMap<T> where T : struct
+    {
+        private const int InitialCapacity = 16;
+        private readonly List<KeyValuePair<int, T>> _dense = new();
+        private int[] _sparse = new int[InitialCapacity];
+
+        private void Grow(int maxValue)
+        {
+            int newLength = maxValue + 1;
```

Comment:
Is there a reason you don't just use a List instead of an array?

Reply:
It could use List too, but the Resize method would still be used and there's no need to append at any point.

Reply:
Using a list would get rid of that logic for doubling the size.

Reply:
Actually List doesn't have a resize method, so the Array seems still simpler to use.

Reply:
It does, there's a setter on `List<T>.Capacity`. (There's also `public int EnsureCapacity(int capacity)` which sets it to provided value or twice current number of items whichever is larger)

It's no biggie, was just curious why you're manually maintaining _sparse but not _dense.

Reply:
I think `Capacity` still just affects the backing array size, so this would need some other function to increase the `Count` of the list. `AddRange` could do that, but it needs an `IEnumerable`, so doesn't seem good for cheaply zero-initializing additional space. By the way, interestingly this data structure wouldn't even need zero-initialization, any random values in unused `_sparse` are still handled fine.